projects/extension/sql/idempotent/001-openai.sql

```diff
@@ -19,196 +19,391 @@ set search_path to pg_catalog, pg_temp
 -- openai_detokenize
 -- decode tokens for a given model back into text
 -- https://github.com/openai/tiktoken/blob/main/README.md
-create or replace function ai.openai_detokenize(model text, tokens int[]) returns text
+create or replace function ai.openai_detokenize(_model text, _tokens int[]) returns text
 as $python$
     #ADD-PYTHON-LIB-DIR
     import tiktoken
-    encoding = tiktoken.encoding_for_model(model)
-    content = encoding.decode(tokens)
+    encoding = tiktoken.encoding_for_model(_model)
+    content = encoding.decode(_tokens)
     return content
 $python$
 language plpython3u strict immutable parallel safe security invoker
 set search_path to pg_catalog, pg_temp
 ;
 
+-------------------------------------------------------------------------------
+-- openai_client_create
+-- create the client and store it in the global dictionary for the session
+CREATE OR REPLACE FUNCTION ai.openai_client_create(
+    _api_key text DEFAULT NULL,
+    _api_key_name text DEFAULT NULL,
+    _organization text DEFAULT NULL,
+    _base_url text DEFAULT NULL,
+    _timeout float8 DEFAULT NULL,
+    _max_retries int DEFAULT NULL,
+    _default_headers jsonb DEFAULT NULL,
+    _default_query jsonb DEFAULT NULL,
+    _http_client jsonb DEFAULT NULL,
+    _strict_response_validation boolean DEFAULT NULL
+) RETURNS void AS $python$
+    #ADD-PYTHON-LIB-DIR
+    import ai.openai
+
+    if 'openai_client' not in GD:
+        GD['openai_client'] = {}
+
+    new_config = ai.openai.prepare_kwargs({
+        'api_key': _api_key,
+        'api_key_name': _api_key_name,
+        'organization': _organization,
+        'base_url': _base_url,
+        'timeout': _timeout,
+        'max_retries': _max_retries,
+        'default_headers': ai.openai.process_json_input(_default_headers),
+        'default_query': ai.openai.process_json_input(_default_query),
+        'http_client': ai.openai.process_json_input(_http_client),
+        '_strict_response_validation': _strict_response_validation
+    })
+
+    if 'config' not in GD['openai_client'] or ai.openai.client_config_changed(GD['openai_client']['config'], new_config):
+        client = ai.openai.make_async_client(plpy, **new_config)
+        GD['openai_client'] = {
+            'client': client,
+            'config': new_config
+        }
+
+$python$ LANGUAGE plpython3u VOLATILE SECURITY DEFINER PARALLEL UNSAFE;
+
+-------------------------------------------------------------------------------
+-- openai_client_destroy
+-- remove the client object stored in the global dictionary for the session
+CREATE OR REPLACE FUNCTION ai.openai_client_destroy() RETURNS void AS $python$
```

Comment:
same question. This seems it should be internal and not exposed to the user.

Reply:
Ditto. Seemed wrong to have a create function without the corresponding cleanup. I'm sure in the majority of use cases this could be ignored.